src/Storages/MaterializedView/RefreshTask.cpp

```diff
@@ -87,19 +87,13 @@ RefreshTask::RefreshTask(
 
         auto zookeeper = context->getZooKeeper();
         String replica_path = coordination.path + "/replicas/" + coordination.replica_name;
-        if (attach)
-        {
-            /// Check that this replica is registered in keeper.
-            if (!zookeeper->exists(replica_path))
-            {
-                LOG_ERROR(log, "Attaching refreshable materialized view {} as read-only because znode {} is missing", view->getStorageID().getFullTableName(), replica_path);
-                coordination.read_only = true;
-            }
-        }
-        else
+        bool replica_path_existed = zookeeper->exists(replica_path);
+
+        /// Create znodes even if it's ATTACH query. This seems weird, possibly incorrect, but
+        /// currently both DatabaseReplicated and DatabaseShared seem to require this behavior.
+        if (!replica_path_existed)
         {
             zookeeper->createAncestors(coordination.path);
-            /// Create coordination znodes if they don't exist. Register this replica, throwing if already exists.
             Coordination::Requests ops;
             ops.emplace_back(zkutil::makeCreateRequest(coordination.path, coordination.root_znode.toString(), zkutil::CreateMode::Persistent, true));
             ops.emplace_back(zkutil::makeCreateRequest(coordination.path + "/replicas", "", zkutil::CreateMode::Persistent, true));
```

Comment:
Shouldn't this node be ephemeral? Otherwise, when we create new replicas in place of the old ones, znodes leak

Reply:
Replicated/SharedMergeTree has a similar znode, and it's not ephemeral; it requires a DROP TABLE to clean up zookeeper (and files). I was just copying that.

Are you saying replicas can be removed without doing a DROP TABLE on them? Hm, it has to be possible, e.g. if the replica had a hardware failure. How does it work? Does DatabaseShared support removing another replica and running table cleanup on its behalf? Or does cloud infra run SYSTEM DROP REPLICA for every replicated table? Or do SharedMergeTree and shared catalog have ephemeral replicas and don't require any cleanup when a replica is removed?

Reply:
> Or does cloud infra run SYSTEM DROP REPLICA for every replicated table?

Yes, we run `SYSTEM DROP REPLICA` for RMT/SMT tables in Make-Before-Break releases. But this command is incompatible with MVs, so old replicas' znodes stay in the keeper even when the replicas are not there anymore.

Reply:
So we should probably add some support of `SYSTEM DROP REPLICA` for RMVs so we can run it in MBB. This would be much better than making the node ephemeral

Reply:
But this should probably be done in a separate PR